bee-network/bee-autopeering/src/task.rs

```diff
@@ -0,0 +1,150 @@
+        self.shutdown_handles.insert(R::NAME.into(), handle);
+
+        self.shutdown_order.push(R::NAME.into(), R::SHUTDOWN_PRIORITY);
+    }
+
+    /// Repeats a command in certain intervals provided a context `T`. Will be shut down gracefully with the rest of
+    /// all spawned tasks by specifying a `name` and a `shutdown_priority`.
+    pub(crate) fn repeat<T, D>(&mut self, cmd: Repeat<T>, mut delay: D, ctx: T, name: &str, shutdown_priority: u8)
+    where
+        T: Send + Sync + 'static,
+        D: Iterator<Item = Duration> + Send + 'static,
+    {
+        let (shutdown_tx, mut shutdown_rx) = oneshot::channel::<()>();
+        self.shutdown_senders.insert(name.into(), shutdown_tx);
+
+        let handle = tokio::spawn(async move {
+            for duration in &mut delay {
+                tokio::select! {
+                    _ = &mut shutdown_rx => break,
+                    _ = time::sleep(duration) => cmd(&ctx),
+                }
+            }
+        });
+        log::trace!("`{}` repeating.", name);
+
+        assert!(!self.shutdown_handles.contains_key(name));
+        self.shutdown_handles.insert(name.into(), handle);
+
+        self.shutdown_order.push(name.into(), shutdown_priority);
+    }
+
+    /// Executes the system shutdown.
+    pub(crate) async fn shutdown(self) {
+        let TaskManager {
+            mut shutdown_order,
+            shutdown_handles: mut runnable_handles,
+            mut shutdown_senders,
+            peerstore,
+            active_peers,
+            replacements,
+        } = self;
+
+        // Send the shutdown signal to all receivers.
+        // TODO: clone necessary?
+        let mut shutdown_order_clone = shutdown_order.clone();
+        while let Some((task_name, _)) = shutdown_order_clone.pop() {
+            // Panic: unwrapping is fine since for every entry in `shutdown_order` there's
+            // a corresponding entry in `shutdown_senders`.
+            let shutdown_tx = shutdown_senders.remove(&task_name).unwrap();
+
+            log::trace!("Shutting down: {}", task_name);
+            shutdown_tx.send(()).expect("error sending shutdown signal");
+        }
+
+        // Wait for all tasks to shutdown down in a certain order and maximum amount of time.
+        time::timeout(SHUTDOWN_TIMEOUT_SECS, async {
+            while let Some((task_name, _)) = shutdown_order.pop() {
+                // Panic: unwrapping is fine, because we are in control of the data.
+                let task_handle = runnable_handles.remove(&task_name).unwrap();
+
+                match task_handle.await {
+                    Ok(_) => {
+                        log::trace!("`{}` stopped.", task_name);
+                    }
+                    Err(e) => {
+                        log::error!("Error shutting down `{}`. Cause: {}", task_name, e);
+                    }
+                }
+            }
+        })
+        .await
+        .expect("error awaiting shutdown");
```

Comment:
Is this happening when the timeout expired ? If yes I wouldn't `expect` but just log an error and proceed. Gives a chance to the rest to gracefully shutdown.

Reply:
Done